[visionOS] Upstream WKSeparatedImageView implementation

etiennesegonzac opened the pull request:
#### 62d83e48a0d38c1382c7b1f7765fa53ad8b3c3b7
<pre>
[visionOS] Upstream WKSeparatedImageView implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=304500">https://bugs.webkit.org/show_bug.cgi?id=304500</a>
&lt;<a href="https://rdar.apple.com/156490053">rdar://156490053</a>&gt;

Reviewed by Richard Robinson.

No logic change, but this feature is finally broken down into multiple
files and obvious Swift warnings / lint issues were addressed.

Tests: LayoutTests/platform/visionos/transforms/separated-image*

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift: Added.
(WKSeparatedImageView.pickViewMode):
(WKSeparatedImageView.startImageAnalysis):
(taskQueue):
(idleTask):
(idleContinuation):
(scoreImage(_:)):
(isIdle):
(enqueue(_:)):
(processQueue):
(CGImage.hasCompatibleDimensions):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift: Added.
(WKSeparatedImageView.startImage3DGeneration):
(tail):
(getTurn() async -&gt; @Sendable (:)):
(ContinuationBox.continuation):
(makeVoidPromise() -&gt; (promise: Task&lt;Void:Never:fulfill:)):
(StoredData.desiredViewingModeSpatial):
(set(for:spatial3DImage:desiredViewingModeSpatial:)):
(updateDesiredViewingMode(for:_:)):
(get(for:)):
(forget(_:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Rendering.swift: Added.
(WKSeparatedImageView.layoutCustomSubtree):
(WKSeparatedImageView.scheduleUpdate):
(WKSeparatedImageView.update):
(WKSeparatedImageView.updateContents):
(WKSeparatedImageView.updatePortal):
(WKSeparatedImageView.requestPortalEntity):
(WKSeparatedImageView.preparePortalEntity):
(WKSeparatedImageView.updateUI):
(WKSeparatedImageView.generationQueueSpinnerView):
(WKSeparatedImageView.desiredViewModeToggleView):
(WKSeparatedImageView.nextDisplayLink):
(ViewMode.description):
(DesiredViewModeButton.isSpatial):
(DesiredViewModeButton.body):
(DisplayLinkTarget.continuation):
(DisplayLinkTarget.handle(_:)):
(CALayer.effectivePointsPerMeter):
(UIView.replaceSubviews(with:)):
(UIView.updateVisibility(_:animate:duration:animationCompletion:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Surface.swift: Added.
(WKSeparatedImageView.processSurface(_:)):
(WKSeparatedImageView.didReceiveImage):
(StreamHasher.outputStream):
(cgImageAsDataWithHash(_:NSString:)):
(makeCGImage(from:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.h: Renamed from Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.h.
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.swift: Added.
(WKSeparatedImageView.viewMode):
(WKSeparatedImageView.cachedViewModeInfo):
(WKSeparatedImageView.desiredViewingModeSpatial):
(WKSeparatedImageView.cgImage):
(WKSeparatedImageView.imageData):
(WKSeparatedImageView.imageHash):
(WKSeparatedImageView.computeHashTask):
(WKSeparatedImageView.pickViewModeTask):
(WKSeparatedImageView.generate3DImageTask):
(WKSeparatedImageView.spatial3DImage):
(WKSeparatedImageView.portalEntity):
(WKSeparatedImageView.contentView):
(WKSeparatedImageView.cornerView):
(WKSeparatedImageView.spinner):
(WKSeparatedImageView.toggleHostingView):
(WKSeparatedImageView.isSeparated):
(WKSeparatedImageView.isAttached):
(WKSeparatedImageView.isRenderingIPC):
(WKSeparatedImageView.logPrefix):
(WKSeparatedImageView.scaleObservation):
(WKSeparatedImageView.setSurface(_:)):
(WKSeparatedImageView.layoutSubviews):
(WKSeparatedImageView.layerClass):
(WKSeparatedImageView.layerSeparatedDidChange):
(WKSeparatedImageView.layerWasCleared):
(WKObservingLayerDelegate.layerSeparatedDidChange):
(WKObservingLayerDelegate.layerWasCleared):
(WKObservingLayer.layerDelegate):
(WKObservingLayer.isSeparated):
(WKObservingLayer.contents):
(WKSeparatedImageView.cleanUp):
(WKSeparatedImageView.webSetSubViews(_:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageViewConstants.swift: Added.
* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.mm: Removed.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
Move the WKContentControlled conformance here since we don&apos;t have an
objective-c implementation anymore.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/305290@main">https://commits.webkit.org/305290@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a8c9a32267c24dc18dcab08b36b79db6c4e5613

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146075 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/90979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/578ba1a6-5289-41ce-9151-d5bd42d205b8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105542 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/90979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bc41a42e-9bb9-4887-8549-5fced0d96432) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86391 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3741aea5-f72b-4875-878f-b4bcb61e22e7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7877 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5633 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6356 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117274 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148784 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10051 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42469 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113941 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10068 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114277 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29031 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7814 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119996 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/64775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10097 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/37971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9828 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/73665 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10038 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9889 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Modules/Internal/WebKitInternal.h:
```diff
@@ -32,8 +32,11 @@
 #import "_WKTextExtractionInternal.h"
+
+#if HAVE(CORE_ANIMATION_SEPARATED_LAYERS)
+#import "WKSeparatedImageView.h"
```
umbrella header files should not have any compile guards like `HAVE(…` ; those guards should just be in their respective header files instead

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    func startImageAnalysis() -> Task<ViewMode, Error> {
+        Task.detached { [weak self] in
```
`Task.detached` should essentially never be used

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift:
```diff
@@ -0,0 +1,223 @@
+
+// Protects access to the taskQueue, requests run sequentially outside of the MainActor.
+actor AnalysisActor {
+    typealias Res = CalculateImageAestheticsScoresRequest.Result?
```
Res -> Result

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift:
```diff
@@ -0,0 +1,223 @@
+
+    static let shared = AnalysisActor()
+
+    private var taskQueue: [() async -> Void] = []
```
```suggestion
    private var taskQueue: [() async -> ()] = []
```

etiennesegonzac replied:
The formatter brings it back to Void

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift:
```diff
@@ -0,0 +1,223 @@
+            var request = CalculateImageAestheticsScoresRequest()
+            let supportedComputeStageDevices = request.supportedComputeStageDevices
+            guard
+                let gpuDevice = supportedComputeStageDevices[.main]?
```
this should be split out into a statement before the guard, so that way trailing closures can be used and then the guard can just be like `guard gpuDevice else { whatever }`

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift:
```diff
@@ -0,0 +1,170 @@
+internal import WebKit_Internal
+
+extension WKSeparatedImageView {
+    func startImage3DGeneration() -> Task<Void, Error> {
```
shouldn’t this be `async` instead of returning a Task?

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift:
```diff
@@ -0,0 +1,170 @@
+            }
+        }
+
+        return Task.detached { [weak self] () -> Void in
```
ditto about avoiding Task.detached

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+private final class ContinuationBox: @unchecked Sendable {
```
what’s up with this? Is it provable safe to be sendable?

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift:
```diff
@@ -0,0 +1,170 @@
+
+#if canImport(RealityFoundation, _version: 387)
+@MainActor
+class ImagePresentationCache {
```
```suggestion
final class ImagePresentationCache {
```

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift:
```diff
@@ -0,0 +1,170 @@
+        self.cache = cache
+    }
+
+    func set(for hash: NSString, spatial3DImage: ImagePresentationComponent.Spatial3DImage, desiredViewingModeSpatial: Bool) {
```
can this (and the below functions) just be a single `subscript` function?

etiennesegonzac replied:
for the follow-up

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Rendering.swift:
```diff
@@ -0,0 +1,348 @@
+        var description: String {
+            switch self {
+            case .portal:
+                return "ImagePresentationComponent"
```
no `return` keyword needed here

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Rendering.swift:
```diff
@@ -0,0 +1,348 @@
+    }
+}
+
+private class DisplayLinkTarget: NSObject {
```
```suggestion
private final class DisplayLinkTarget: NSObject {
```

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Rendering.swift:
```diff
@@ -0,0 +1,348 @@
+    }
+}
+
+extension UIView {
```
should probably be in a `UIView+Extras.swift` file

etiennesegonzac replied:
But then I'd have to remove the fileprivate? these felt a bit too specific to be shared

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.swift:
```diff
@@ -0,0 +1,234 @@
+    @nonobjc
+    var viewMode: ViewMode
+    @nonobjc
+    var cachedViewModeInfo: (NSString, ViewMode)?
```
why NSString and not String?

etiennesegonzac replied:
it eventually gets passed to the NSCache

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.swift:
```diff
@@ -0,0 +1,234 @@
+    @nonobjc
+    var toggleHostingView: UIView?
+
+    @nonobjc
+    var isSeparated: Bool = false
+    @nonobjc
+    var isAttached: Bool = false
+    @nonobjc
+    var isRenderingIPC: Bool = false
```
why do these need to be nonobjc?

rr-codes commented on Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.swift:
```diff
@@ -0,0 +1,234 @@
+
+// MARK: - Extensions
+
+extension Logger {
```
should probably be in a Logger+Extras.swift file

pxlcoder reviewed (commented):
Ideally an upstreaming patch should just move code from one place to another. Any refactoring / actual code changes would be done in subsequent PRs.

etiennesegonzac committed:
[visionOS] Upstream WKSeparatedImageView implementation
https://bugs.webkit.org/show_bug.cgi?id=304500
<rdar://156490053>

Reviewed by Richard Robinson.

No logic change, but this feature is finally broken down into multiple
files and obvious Swift warnings / lint issues were addressed.

Tests: LayoutTests/platform/visionos/transforms/separated-image*

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Analysis.swift: Added.
(WKSeparatedImageView.pickViewMode):
(WKSeparatedImageView.startImageAnalysis):
(taskQueue):
(idleTask):
(idleContinuation):
(scoreImage(_:)):
(isIdle):
(enqueue(_:)):
(processQueue):
(CGImage.hasCompatibleDimensions):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Generation.swift: Added.
(WKSeparatedImageView.startImage3DGeneration):
(tail):
(getTurn() async -> @Sendable (:)):
(ContinuationBox.continuation):
(makeVoidPromise() -> (promise: Task<Void:Never:fulfill:)):
(StoredData.desiredViewingModeSpatial):
(set(for:spatial3DImage:desiredViewingModeSpatial:)):
(updateDesiredViewingMode(for:_:)):
(get(for:)):
(forget(_:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Rendering.swift: Added.
(WKSeparatedImageView.layoutCustomSubtree):
(WKSeparatedImageView.scheduleUpdate):
(WKSeparatedImageView.update):
(WKSeparatedImageView.updateContents):
(WKSeparatedImageView.updatePortal):
(WKSeparatedImageView.requestPortalEntity):
(WKSeparatedImageView.preparePortalEntity):
(WKSeparatedImageView.updateUI):
(WKSeparatedImageView.generationQueueSpinnerView):
(WKSeparatedImageView.desiredViewModeToggleView):
(WKSeparatedImageView.nextDisplayLink):
(ViewMode.description):
(DesiredViewModeButton.isSpatial):
(DesiredViewModeButton.body):
(DisplayLinkTarget.continuation):
(DisplayLinkTarget.handle(_:)):
(CALayer.effectivePointsPerMeter):
(UIView.replaceSubviews(with:)):
(UIView.updateVisibility(_:animate:duration:animationCompletion:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView+Surface.swift: Added.
(WKSeparatedImageView.processSurface(_:)):
(WKSeparatedImageView.didReceiveImage):
(StreamHasher.outputStream):
(cgImageAsDataWithHash(_:NSString:)):
(makeCGImage(from:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.h: Renamed from Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.h.
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageView.swift: Added.
(WKSeparatedImageView.viewMode):
(WKSeparatedImageView.cachedViewModeInfo):
(WKSeparatedImageView.desiredViewingModeSpatial):
(WKSeparatedImageView.cgImage):
(WKSeparatedImageView.imageData):
(WKSeparatedImageView.imageHash):
(WKSeparatedImageView.computeHashTask):
(WKSeparatedImageView.pickViewModeTask):
(WKSeparatedImageView.generate3DImageTask):
(WKSeparatedImageView.spatial3DImage):
(WKSeparatedImageView.portalEntity):
(WKSeparatedImageView.contentView):
(WKSeparatedImageView.cornerView):
(WKSeparatedImageView.spinner):
(WKSeparatedImageView.toggleHostingView):
(WKSeparatedImageView.isSeparated):
(WKSeparatedImageView.isAttached):
(WKSeparatedImageView.isRenderingIPC):
(WKSeparatedImageView.logPrefix):
(WKSeparatedImageView.scaleObservation):
(WKSeparatedImageView.setSurface(_:)):
(WKSeparatedImageView.layoutSubviews):
(WKSeparatedImageView.layerClass):
(WKSeparatedImageView.layerSeparatedDidChange):
(WKSeparatedImageView.layerWasCleared):
(WKObservingLayerDelegate.layerSeparatedDidChange):
(WKObservingLayerDelegate.layerWasCleared):
(WKObservingLayer.layerDelegate):
(WKObservingLayer.isSeparated):
(WKObservingLayer.contents):
(WKSeparatedImageView.cleanUp):
(WKSeparatedImageView.webSetSubViews(_:)):
* Source/WebKit/UIProcess/Cocoa/Separated/WKSeparatedImageViewConstants.swift: Added.
* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.mm: Removed.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
Move the WKContentControlled conformance here since we don't have an
objective-c implementation anymore.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/305290@main
